src/qb4-packw/kr-scalar.c.in

```diff
@@ -0,0 +1,200 @@
+// Copyright 2023 Google LLC
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$assert NR > 1
+$assert KR > 1
+#include <assert.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <math.h>
+
+#include "xnnpack/packw.h"
+
+void xnn_qb4_packw_gemm_goi_ukernel_x${NR}c${KR}__scalar(
+  size_t g,
+  size_t nc,
+  size_t kc,
+  size_t nr,
+  size_t kr,
+  size_t sr,
+  size_t bl,
+  const uint8_t* weights,
+  const int32_t* bias,
+  const void* scale,
+  int8_t* packed_weights,
+  size_t extra_bytes_bl,
+  size_t extra_bytes_n,
+  const void* params)
+{
+  assert(g != 0);
+  assert(nc != 0);
+  assert(kc != 0);
+  assert(nr == ${NR});
+  assert(kr == ${KR});
+  assert(sr == 1);
+  assert(weights != NULL);
+  assert(packed_weights != NULL);
+  assert(extra_bytes_bl == nr * sizeof(uint16_t));
+  assert(extra_bytes_n == nr * sizeof(float));
+  assert(params != NULL);
+  assert(kc % bl == 0);
+  size_t num_blocks = kc / bl;
+
+  int8_t* out = (int8_t*) packed_weights;
+  const int32_t* b = (const int32_t*) bias;
+  const uint32_t izp = (uint32_t) (((const struct xnn_qs8_qc4w_packing_params*) params)->input_zero_point + 0);
+
+  do {
+    // NC main loop multiple of ${NR}
+    const uint8_t* w0 = (const uint8_t*) weights;
+    const uint16_t* s0 = (const uint16_t*) scale;
+    size_t n = nc;
+    for (;n >= ${NR}; n -= ${NR}) {
+        float* packed_k_scaled_sum = (float*) out;
+        $for N in range(NR):
+            ((float*) out)[${N}] = 0;
+        out += ${NR} * sizeof(float);
+
+        // KC/2 bytes is KC Nibbles
+        $for N in range(1, NR):
+            const uint8_t* w${N} = w${N-1} + (kc >> 1);
```

Comment:
this kernel does not support odd KC?
how do we ensure convolutions dont call this when kc is odd?

for x4-pack scalar kernel it still supports any KC, inefficiently
then from avxvnni I check if KC is odd and call the scalar kernel to handle it

Reply:
This kernel isn't used for convolutions, and only for fully connected layers. 

A constraint applied on blockwise quantization is that kc must be divisible by blocksize, and block size is a multiple 32.

Reply:
code pointer? Let's make sure they are not just asserts and we fail with proper error message if not doing that already.